### PR TITLE
reconciler: 2 phase reconnects and tests

### PR DESCRIPTION
**Pull request by DerekStrickland**
Not ready for review

This PR changes the generic scheduler reconcile logic to support zero-downtime reconnects.

The initial implementation for disconnected clients stopped replacement allocations before the reconnecting node synced its alloc state with the server. If the alloc failed while the node was disconnected, this could lead to a scenario where the replacement was stopped, but the reconnecting alloc was not healthy, and the cluster ends up underprovisioned.

This PR changes the logic so that allocs are identified as reconnecting after the reconnecting client sends its alloc updates to the server.  To make this work, I made the following changes.

- `Node.UpdateAlloc`
  - Looks up the server-side alloc state immediately instead of after checking if the incoming client alloc is terminal.
  - Previously, the incoming alloc update would be skipped if it was in a terminal state on the client. 
  - Now it will only be skipped if it is in a terminal state at the client and not unknown at the server.
  - Any incoming alloc updates that are unknown at the sever will now trigger an eval with `EvalTriggerReconnect`.
- I added an`Expired`function to `Allocation` that uses `AllocStates` to determine if an unknown allocation has expired.
- I updated the reconciler to append an `AllocState` entry when creating the unknown alloc update.
-  I added a`Reconnected`function to `Allocation` that uses `AllocStates` plus `TaskStates` to determine if an allocation has reconnected and whether it expired before reconnecting.
- I updated the `filterByTainted` function to:
  - Filter allocs on disconnected nodes into `disconnecting`, `lost`, or `ignore` based on client status and expiration 
 state.
   - Filter allocs on reconnecting nodes into `reconnecting`, `lost`, or `ignore` based on their alloc and task states.
- I updated the `filterByResheduable` function to also be able to handle `disconnecting` allocations.
- I updated the reconciler to:
  - Incorporate the `ignore` set now returned by `filterByTainted` into `DesiredChanges`
  - Consider job version/create_index in multiple areas.
- I updated the `allocWatcher` to also trigger when a previous allocation transitions to unknown

### CODE REVIEW

File: scheduler/reconcile.go
Comment (DerekStrickland) [inline scheduler/reconcile.go:747]: This looks wrong. I'm taking a look at it now, but don't want to hold up the PR review.

File: scheduler/reconcile.go
Comment (DerekStrickland) [inline scheduler/reconcile.go:1038]: Should these just be != instead of < ?  Is a rollback possible?  Same question for all these checks.

Comment (tgross) [reply scheduler/reconcile.go:1038]: I definitely _shouldn't_ be possible to roll back a version (a rolled-back deployment always creates a new version), but having the `<` assertion seems like it gives us some guard rails for sure. I'd keep it like this.

File: scheduler/reconcile.go
Comment (tgross) [reply scheduler/reconcile.go:747]: Adding "reschedule now" to the stop set seems to make sense. We need to server-stop allocs that we're replacing (which might only be client-failed), right?

**Review by tgross — COMMENTED** [review]
Overall this is looking good @DerekStrickland. I've left some suggestions on the tests, as making those legible will be important for maintenance. And then you've got a couple of open questions that we should resolve before it's ready to merge.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:747]: I remember why this is here now, but I didn't comment on it well originally. I reviewed the code and am adding the following comments to clarify. I'm also renaming `stop` to `failed` so that it clarifies the intent of this function.

```go
// Disconnecting allocs are not failing, but are included in rescheduleNow.
// Create a new set that only includes the actual failures and compute
// replacements based off that.
```

File: scheduler/reconcile_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_test.go:5358]: Great call out. Will be fixed in the next push.

File: scheduler/reconcile_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_test.go:5403]: Updated in next push.

File: scheduler/reconcile_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_test.go:5460]: Yes!  Nice catch.  Resolved in next push.

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:10031]: It is trying to avoid multiple calls to `LastUnknown` which iterates over `AllocStates`.  `Expired` and `Reconnected` both need to do this. It's also trying a way of checking the expiration without checking the `ClientStatus`. You just spotted a bug in Reconnected though. It can't call `Expired` and be valid.  I'll rework this.

Comment (DerekStrickland) [reply nomad/structs/structs.go:10031]: I think this was a refactoring that went sideways.  I've made it much simpler. Sorry for the oops. I'm leaving this comment open so you can review it in the next push.

File: nomad/structs/structs_test.go
Comment (DerekStrickland) [reply nomad/structs/structs_test.go:5089]: Will be fixed in the next push.

File: nomad/structs/structs_test.go
Comment (DerekStrickland) [reply nomad/structs/structs_test.go:5093]: Good point. I'll force a different timezone in the next push.

File: scheduler/reconcile_util.go
Comment (DerekStrickland) [reply scheduler/reconcile_util.go:218]: Fixed in next push.

File: scheduler/reconcile_util.go
Comment (DerekStrickland) [reply scheduler/reconcile_util.go:351]: Fixed in next push. Leaving this comment open so you can review it.

File: scheduler/reconcile_util_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_util_test.go:533]: This is for logging what's different. The next line handles the require. IIRC I did this because `require`'s output for map diffs is pretty right scrolly and lack-luster. I can change though if you want.

File: scheduler/reconcile_util_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_util_test.go:528]: Updated in the next push.

File: scheduler/reconcile_util_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_util_test.go:544]: Fixed in next push.

File: scheduler/reconcile_util_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_util_test.go:575]: Copy pasta bug.  Should have passed nil.

File: client/allocrunner/alloc_runner.go
Comment (DerekStrickland) [reply client/allocrunner/alloc_runner.go:1274]: Fixed in next push.

File: nomad/structs/structs.go
Comment (tgross) [reply nomad/structs/structs.go]: Much nicer!

**Review by tgross — APPROVED** [review]
LGTM!

File: scheduler/reconcile_util_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_util_test.go:87]: Added comments explaining.

**Pull request merged**
